### PR TITLE
Property Refactoring

### PR DESCRIPTION
**Pull request by bhargav**
- Decoupling Property and LBJava's classifier. 
- Property still interfaces using `FeatureVector`

Note: Some unit tests currently fail because the models need to be re-trained. Will train models after the PR is reviewed at least once.

### CODE REVIEW

**Comment by kordjamshidi** [comment]
"Decoupling Property and LBJava's classifier." is that something we wanted to do?

**Comment by bhargav** [comment]
With this PR, Property type still uses FeatureVector and Feature types from LBJava. 

The main idea was that moving away from LBJava types would help integrating other algorithms -- SL Integration, Weka, etc. We should have our own data structures/interfaces for Properties and Classifiers. We can discuss this further sometime. This PR still uses LBJava's types and removes duplicated code from Property sub-classes.

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/Learnable.scala [hunk 1]
```diff
@@ -54,7 +54,7 @@ abstract class Learnable[T <: AnyRef](val node: Node[T], val parameters: Paramet
   def combinedProperties = if (label != null) new CombinedDiscreteProperty[T](this.feature.filterNot(_.name == label.name))
   else new CombinedDiscreteProperty[T](this.feature)
 
-  def lbpFeatures = combinedProperties.classifier
+  def lbpFeatures = Property.convertToClassifier(combinedProperties)
```
Comment (danyaljj) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/Learnable.scala]: 👍

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/Learnable.scala [hunk 2]
```diff
@@ -102,7 +102,7 @@ abstract class Learnable[T <: AnyRef](val node: Node[T], val parameters: Paramet
 
   private def setLabeler(): Unit = {
     if (label != null) {
-      val oracle = Property.entitiesToLBJFeature(label)
+      val oracle = Property.convertToClassifier(label)
```
Comment (danyaljj) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/classifier/Learnable.scala]: 👍

File: saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/property/ClassifierContainsInLBP.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  * Developed by: The Cognitive Computations Group, University of Illinois at Urbana-Champaign
+  * http://cogcomp.cs.illinois.edu/
+  */
+package edu.illinois.cs.cogcomp.saul.datamodel.property
+
+import java.util
+
+import edu.illinois.cs.cogcomp.lbjava.classify.{ Classifier, FeatureVector }
+
+private[property] class ClassifierContainsInLBP[T](val property: Property[T]) extends Classifier {
```
Comment (danyaljj) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/property/ClassifierContainsInLBP.scala]: The naming sounds weird to me. Why "contains"?

Comment (danyaljj) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/property/ClassifierContainsInLBP.scala]: Just noticed that there was another file with this name. Still, maybe we can choose a better name? (I'm brain dead now)

Comment (bhargav) [inline saul-core/src/main/scala/edu/illinois/cs/cogcomp/saul/datamodel/property/ClassifierContainsInLBP.scala]: I have moved the older one to a this location. Also I am open to renaming it. 😛

**Comment by danyaljj** [comment]
I like the changes. It moves us in the direction of decoupling LBJ classifier and properties/sensors. This will let us to make the implementation of properties more efficient and cleaner. Meanwhile later we can decide whether we want to use LBJava's feature extraction or have our own. 

Semaphore is unhappy; SRL probably needs retraining? 

Good to merge by me when Semaphore and Parisa are happy.

File: saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala [hunk 4]
```diff
@@ -82,9 +82,9 @@ class ModelsTest extends FlatSpec with Matchers {
     results.perLabel.foreach {
       result =>
         result.label match {
-          case "A0" => result.f1 should be(0.95 +- 0.05)
-          case "A1" => result.f1 should be(0.95 +- 0.05)
-          case "A2" => result.f1 should be(0.8 +- 0.03)
+          case "A0" => result.f1 should be >= 0.9
+          case "A1" => result.f1 should be >= 0.9
+          case "A2" => result.f1 should be >= 0.8
```
Comment (danyaljj) [inline saul-examples/src/test/scala/edu/illinois/cs/cogcomp/saulexamples/nlp/SemanticRoleLabeling/ModelsTest.scala]: Sorry for being nitpicky. It bothers be when the tests are not upper-bounded. Could you make it more strict?

**Comment by danyaljj** [comment]
One minor comment about upper-bounds in the unit tests. Otherwise looks good.

**Comment by danyaljj** [comment]
@kordjamshidi do you have any comments on this PR? If not, I think we can merge this.

**Comment by kordjamshidi** [comment]
Give me time until tomorrow, will look at it.

**Comment by kordjamshidi** [comment]
I was looking into this. It would be good if we can also discuss this tomorrow before merging.

**Comment by bhargav** [comment]
Sure. That works.

**Comment by danyaljj** [comment]
Update it with `IllinoisCogComp/master`?

**Comment by kordjamshidi** [comment]
@bhargav : I think we decided on some documentation before merging this, I just noticed you have merged this.

**Comment by kordjamshidi** [comment]
Sorry, @danyaljj.

**Comment by danyaljj** [comment]
Bhargavi added some comments about the removal of the classifier function before merging. What other documentation did you have in mind?

**Comment by kordjamshidi** [comment]
ah sorry, so, I could not find that commit in the list then.

**Pull request merged**